com.sun.xml.ws.model.Injector incompatible with Java 11

With OpenJDK 11.0.2 and JAX-WS 2.3.2, calling ``javax.xml.ws.Service.getPort(QName, Class<T>)`` (e.g. from a service implementation) leads to a ``java.lang.NoSuchMethodException`` (after fixing issue #54 by the workaround mentioned there). This exception is thrown in com.sun.xml.ws.model.Injector, as the method ``sun.misc.Unsafe.defineClass(...)`` no longer exists in Java 11 after being deprecated in Java 9 (see also https://bugs.openjdk.java.net/browse/JDK-8193033):

```
java.security.PrivilegedActionException: java.lang.NoSuchMethodException: sun.misc.Unsafe.defineClass(java.lang.String, [B, int, int, java.lang.ClassLoader, java.security.ProtectionDomain)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.Injector.<clinit>(Injector.java:75)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.WrapperBeanGenerator.createRequestWrapperBean(WrapperBeanGenerator.java:265)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.RuntimeModeler.getRequestWrapperClass(RuntimeModeler.java:345)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.RuntimeModeler.processDocWrappedMethod(RuntimeModeler.java:771)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.RuntimeModeler.processMethod(RuntimeModeler.java:701)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.RuntimeModeler.processClass(RuntimeModeler.java:461)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.RuntimeModeler.buildRuntimeModel(RuntimeModeler.java:290)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.db.DatabindingImpl.<init>(DatabindingImpl.java:70)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.db.DatabindingProviderImpl.create(DatabindingProviderImpl.java:44)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.db.DatabindingProviderImpl.create(DatabindingProviderImpl.java:28)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.db.DatabindingFactoryImpl.createRuntime(DatabindingFactoryImpl.java:90)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.client.WSServiceDelegate.buildRuntimeModel(WSServiceDelegate.java:852)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.client.WSServiceDelegate.createSEIPortInfo(WSServiceDelegate.java:869)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.client.WSServiceDelegate.addSEI(WSServiceDelegate.java:832)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.client.WSServiceDelegate.getPort(WSServiceDelegate.java:421)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.client.WSServiceDelegate.getPort(WSServiceDelegate.java:390)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.client.WSServiceDelegate.getPort(WSServiceDelegate.java:372)
	at java.xml.ws/javax.xml.ws.Service.getPort(Service.java:139)
	[...]
Caused by: java.lang.NoSuchMethodException: sun.misc.Unsafe.defineClass(java.lang.String, [B, int, int, java.lang.ClassLoader, java.security.ProtectionDomain)
	at java.base/java.lang.Class.getMethod(Class.java:2109)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.Injector$3.run(Injector.java:79)
	at com.sun.xml.ws.jaxws/com.sun.xml.ws.model.Injector$3.run(Injector.java:75)
	... 60 more
```

A possible workaround is to start your application with the ``--add-opens java.base/java.lang=com.sun.xml.ws.jaxws`` flag. In this case, the ``defineClass(...)`` method of ``java.lang.ClassLoader`` is used instead than the one from ``sun.misc.Unsafe``.

However, I would expect the JAX-WS library to work without any additional ``--add-opens`` statements.

--- a/jaxws-ri/bundles/jaxws-rt/pom.xml
+++ b/jaxws-ri/bundles/jaxws-rt/pom.xml
@@ -288,7 +288,6 @@
                                 <Import-Package>
                                     javax.xml.catalog;resolution:=optional,
                                     sun.misc;resolution:=optional,
-                                    jdk.internal.misc;resolution:=optional,
                                     com.sun.xml.ws.assembler.metro.jaxws;resolution:=optional,
                                     *
                                 </Import-Package>

--- a/jaxws-ri/bundles/jaxws-rt/src/main/java/module-info.java
+++ b/jaxws-ri/bundles/jaxws-rt/src/main/java/module-info.java
@@ -28,6 +28,7 @@ module com.sun.xml.ws {
     requires java.management;
     requires transitive java.xml;
     requires jdk.httpserver;
+    requires jdk.unsupported;
 
     requires transitive jakarta.activation;
     requires transitive jakarta.annotation;

--- a/jaxws-ri/extras/eclipselink_jaxb/pom.xml
+++ b/jaxws-ri/extras/eclipselink_jaxb/pom.xml
@@ -102,9 +102,6 @@
                         <additionalClasspathElement>${project.build.outputDirectory}</additionalClasspathElement>
                     </additionalClasspathElements>
                     <forkCount>1</forkCount>
-                    <argLine>
-                        --add-opens java.base/java.lang=ALL-UNNAMED
-                    </argLine>
                 </configuration>
             </plugin>
             <plugin>

--- a/jaxws-ri/pom.xml
+++ b/jaxws-ri/pom.xml
@@ -277,11 +277,15 @@
                         </execution>
                     </executions>
                 </plugin>
-
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-surefire-plugin</artifactId>
                     <version>3.0.0-M4</version>
+                </plugin>
+                <plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-failsafe-plugin</artifactId>
+                    <version>3.0.0-M5</version>
                 </plugin>
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>

--- a/jaxws-ri/runtime/rt/pom.xml
+++ b/jaxws-ri/runtime/rt/pom.xml
@@ -240,6 +240,8 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
                 <configuration>
+                    <!-- tests are run using failsafe to properly
+                         use MR jar produced here -->
                     <skip>true</skip>
                 </configuration>
             </plugin>

--- a/jaxws-ri/runtime/rt/pom.xml
+++ b/jaxws-ri/runtime/rt/pom.xml
@@ -176,7 +176,7 @@
                         </configuration>
                     </execution>
                     <execution>
-                        <id>default-compile-mr</id>
+                        <id>default-compile-mr-9</id>
                         <goals>
                             <goal>compile</goal>
                         </goals>
@@ -240,6 +240,16 @@
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
                 <configuration>
+                    <skip>true</skip>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <configuration>
+                    <includes>
+                        <include>**/*Test.java</include>
+                    </includes>
                     <!-- exclude APIs as implementations are needed on the path -->
                     <classpathDependencyExcludes>jakarta.mail:jakarta.mail-api</classpathDependencyExcludes>
                     <systemProperties>
@@ -247,6 +257,14 @@
                     </systemProperties>
                     <!-- SAAJMessageWrapperTest/com.sun.xml.ws.model.Injector -->
                 </configuration>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                    </execution>
+                </executions>
             </plugin>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>

--- a/jaxws-ri/runtime/rt/src/main/java-mr/9/com/sun/xml/ws/model/InjectorHelper.java
+++ b/jaxws-ri/runtime/rt/src/main/java-mr/9/com/sun/xml/ws/model/InjectorHelper.java
@@ -1,0 +1,103 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Distribution License v. 1.0, which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+package com.sun.xml.ws.model;
+
+import java.lang.reflect.AccessibleObject;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import sun.misc.Unsafe;
+
+final class InjectorHelper {
+
+    private static Unsafe UNSAFE;
+    private static long OFFSET = -1;
+
+    static {
+        UNSAFE = getUnsafe();
+        if (UNSAFE != null) {
+            Field f = AccessController.doPrivileged(new PrivilegedAction<Field>() {
+                @Override
+                public Field run() {
+                    try {
+                        return AccessibleObject.class.getDeclaredField("override");
+                    } catch (NoSuchFieldException | SecurityException ex) {
+                        return null;
+                    }
+                }
+            });
+            if (f == null) {
+                try {
+                    f = dummy.class.getDeclaredField("override");
+                } catch (NoSuchFieldException ignore) {
+                }
+            }
+            OFFSET = UNSAFE.objectFieldOffset(f);
+        }
+    }
+
+
+    private InjectorHelper() {
+    }
+
+    static Method getMethod(final Class<?> c, final String methodname, final Class<?>... params) {
+        try {
+            Method m = c.getDeclaredMethod(methodname, params);
+            if (!m.trySetAccessible()) {
+                setAccessible(m);
+            }
+            return m;
+        } catch (NoSuchMethodException e) {
+            // impossible
+            throw new NoSuchMethodError(e.getMessage());
+        }
+    }
+
+    static class dummy {
+
+        boolean override;
+        Object other;
+    }
+
+    private static <T extends AccessibleObject> T setAccessible(T accessor) {
+        if (OFFSET != -1) {
+            UNSAFE.putBoolean(accessor, OFFSET, true);
+        } else {
+            accessor.setAccessible(true); // this may fail
+        }
+        return accessor;
+    }
+
+    private static Unsafe getUnsafe() {
+        try {
+            Field f = AccessController.doPrivileged(new PrivilegedAction<Field>() {
+                @Override
+                public Field run() {
+                    try {
+                        return Unsafe.class.getDeclaredField("theUnsafe");
+                    } catch (NoSuchFieldException | SecurityException ex) {
+                        Logger.getLogger(InjectorHelper.class.getName()).log(Level.SEVERE, null, ex);
+                        return null;
+                    }
+                }
+            });
+            if (f != null) {
+                f.setAccessible(true);
+                return (Unsafe) f.get(null);
+            }
+        } catch (Throwable t) {
+        }
+        return null;
+    }
+
+}

--- a/jaxws-ri/runtime/rt/src/main/java-mr/9/com/sun/xml/ws/model/InjectorHelper.java
+++ b/jaxws-ri/runtime/rt/src/main/java-mr/9/com/sun/xml/ws/model/InjectorHelper.java
@@ -86,16 +86,19 @@ final class InjectorHelper {
                     try {
                         return Unsafe.class.getDeclaredField("theUnsafe");
                     } catch (NoSuchFieldException | SecurityException ex) {
-                        Logger.getLogger(InjectorHelper.class.getName()).log(Level.SEVERE, null, ex);
+                        Logger.getLogger(InjectorHelper.class.getName()).log(Level.WARNING, null, ex);
                         return null;
                     }
                 }
             });
             if (f != null) {
-                f.setAccessible(true);
+                if (!f.trySetAccessible()) {
+                    setAccessible(f);
+                }
                 return (Unsafe) f.get(null);
             }
         } catch (Throwable t) {
+            Logger.getLogger(InjectorHelper.class.getName()).log(Level.WARNING, null, t);
         }
         return null;
     }

--- a/jaxws-ri/runtime/rt/src/main/java/com/sun/xml/ws/model/Injector.java
+++ b/jaxws-ri/runtime/rt/src/main/java/com/sun/xml/ws/model/Injector.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2008, 2020 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2008, 2021 Oracle and/or its affiliates. All rights reserved.
  *
  * This program and the accompanying materials are made available under the
  * terms of the Eclipse Distribution License v. 1.0, which is available at
@@ -10,15 +10,10 @@
 
 package com.sun.xml.ws.model;
 
-import java.lang.reflect.Field;
 import jakarta.xml.ws.WebServiceException;
-import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
-import java.net.URL;
 import java.security.AccessController;
 import java.security.PrivilegedAction;
-import java.security.PrivilegedActionException;
-import java.security.PrivilegedExceptionAction;
 import java.security.ProtectionDomain;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -34,77 +29,22 @@ final class Injector {
     private static final Logger LOGGER = Logger.getLogger(Injector.class.getName());
 
     private static Method defineClass;
-    private static Method resolveClass;
-    private static Method getPackage;
-    private static Method definePackage;
-    private static Object U;
 
     static {
         try {
-            Method[] m = AccessController.doPrivileged(
-                    new PrivilegedAction<Method[]>() {
+            defineClass = AccessController.doPrivileged(
+                    new PrivilegedAction<Method>() {
                 @Override
-                public Method[] run() {
-                    return new Method[]{
-                        getMethod(ClassLoader.class, "defineClass", String.class, byte[].class, Integer.TYPE, Integer.TYPE),
-                        getMethod(ClassLoader.class, "resolveClass", Class.class),
-                        getMethod(ClassLoader.class, "getPackage", String.class),
-                        getMethod(ClassLoader.class, "definePackage",
-                            String.class, String.class, String.class, String.class,
-                            String.class, String.class, String.class, URL.class)
-                    };
+                public Method run() {
+                    return InjectorHelper.getMethod(ClassLoader.class, "defineClass",
+                            String.class, byte[].class, Integer.TYPE, Integer.TYPE, ProtectionDomain.class);
                 }
-            }
-            );
-            defineClass = m[0];
-            resolveClass = m[1];
-            getPackage = m[2];
-            definePackage = m[3];
+            });
 
         } catch (Throwable t) {
-            try {
-                U = AccessController.doPrivileged(new PrivilegedExceptionAction() {
-                    @Override
-                    public Object run() throws Exception {
-                        Class u = Class.forName("jdk.internal.misc.Unsafe");
-                        Field theUnsafe = u.getDeclaredField("theUnsafe");
-                        theUnsafe.setAccessible(true);
-                        return theUnsafe.get(null);
-                    }
-                });
-                defineClass = AccessController.doPrivileged(new PrivilegedExceptionAction<Method>() {
-                    @Override
-                    public Method run() throws Exception {
-                        try {
-                            return U.getClass().getMethod("defineClass",
-                                    new Class[]{String.class,
-                                        byte[].class,
-                                        Integer.TYPE,
-                                        Integer.TYPE,
-                                        ClassLoader.class,
-                                        ProtectionDomain.class});
-                        } catch (NoSuchMethodException | SecurityException ex) {
-                            throw ex;
-                        }
-                    }
-                });
-            } catch (SecurityException | PrivilegedActionException ex) {
-                Logger.getLogger(Injector.class.getName()).log(Level.SEVERE, null, ex);
-                WebServiceException we = new WebServiceException(ex);
-                we.addSuppressed(t);
-                throw we;
-            }
-        }
-    }
-
-    private static Method getMethod(final Class<?> c, final String methodname, final Class<?>... params) {
-        try {
-            Method m = c.getDeclaredMethod(methodname, params);
-            m.setAccessible(true);
-            return m;
-        } catch (NoSuchMethodException e) {
-            // impossible
-            throw new NoSuchMethodError(e.getMessage());
+            Logger.getLogger(Injector.class.getName()).log(Level.SEVERE, null, t);
+            WebServiceException we = new WebServiceException(t);
+            throw we;
         }
     }
 
@@ -117,23 +57,9 @@ final class Injector {
             // nothing to do
         }
         try {
-            if (definePackage == null) {
-                return (Class) defineClass.invoke(U, className.replace('/', '.'), image, 0, image.length, cl, Injector.class.getProtectionDomain());
-            }
-            int packIndex = className.lastIndexOf('.');
-            if (packIndex != -1) {
-                String pkgname = className.substring(0, packIndex);
-                // Check if package already loaded.
-                Package pkg = (Package) getPackage.invoke(cl, pkgname);
-                if (pkg == null) {
-                    definePackage.invoke(cl, pkgname, null, null, null, null, null, null, null);
-                }
-            }
-
-            Class c = (Class) defineClass.invoke(cl, className.replace('/', '.'), image, 0, image.length);
-            resolveClass.invoke(cl, c);
-            return c;
-        } catch (IllegalAccessException | InvocationTargetException e) {
+                return (Class) defineClass.invoke(cl,
+                        className.replace('/', '.'), image, 0, image.length, Injector.class.getProtectionDomain());
+        } catch (ReflectiveOperationException e) {
             if (LOGGER.isLoggable(Level.FINE)) {
                 LOGGER.log(Level.FINE, "Unable to inject " + className, e);
             }
@@ -142,4 +68,3 @@ final class Injector {
     }
 
 }
-

--- a/jaxws-ri/runtime/rt/src/main/java/com/sun/xml/ws/model/InjectorHelper.java
+++ b/jaxws-ri/runtime/rt/src/main/java/com/sun/xml/ws/model/InjectorHelper.java
@@ -1,0 +1,30 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates. All rights reserved.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Distribution License v. 1.0, which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+package com.sun.xml.ws.model;
+
+import java.lang.reflect.Method;
+
+final class InjectorHelper {
+
+    private InjectorHelper() {}
+
+    static Method getMethod(final Class<?> c, final String methodname, final Class<?>... params) {
+        try {
+            Method m = c.getDeclaredMethod(methodname, params);
+            m.setAccessible(true);
+            return m;
+        } catch (NoSuchMethodException e) {
+            // impossible
+            throw new NoSuchMethodError(e.getMessage());
+        }
+    }
+
+}

--- a/jaxws-ri/runtime/rt/src/main/java/module-info.java
+++ b/jaxws-ri/runtime/rt/src/main/java/module-info.java
@@ -28,6 +28,7 @@ module com.sun.xml.ws.rt {
     requires java.management;
     requires transitive java.xml;
     requires jdk.httpserver;
+    requires jdk.unsupported;
 
     requires transitive jakarta.activation;
     requires transitive jakarta.annotation;

--- a/jaxws-ri/tests/pom.xml
+++ b/jaxws-ri/tests/pom.xml
@@ -38,6 +38,7 @@
         <ws.test>${project.basedir}/testcases</ws.test>
         <!-- extra JVM options for JVM forked by test harness, ie proxy, tests' debugging-->
         <ws.jvmOpts></ws.jvmOpts>
+        <ws.permGenJvmOpts></ws.permGenJvmOpts>
     </properties>
 
     <build>
@@ -63,9 +64,7 @@
                     <transport>${ws.transport}</transport>
                     <extraVmArgs>${ws.jvmOpts}</extraVmArgs>
                     <vmArgs>
-                        <vmArg>-Xms256m -Xmx512m</vmArg>
-                        <vmArg>--add-opens java.base/java.lang=ALL-UNNAMED</vmArg>
-                        <vmArg>--add-opens java.base/jdk.internal.misc=ALL-UNNAMED</vmArg>
+                        <vmArg>-Xms256m -Xmx512m ${ws.permGenJvmOpts}</vmArg>
                         <vmArg>--add-opens java.base/java.util=ALL-UNNAMED</vmArg>
                         <vmArg>--add-opens java.xml/com.sun.org.apache.xalan.internal.xsltc.trax=ALL-UNNAMED</vmArg>
                         <vmArg>--add-opens java.xml/com.sun.org.apache.xerces.internal.jaxp=ALL-UNNAMED</vmArg>

